listenerutil/forwarded_for.go

```diff
@@ -0,0 +1,192 @@
+			return
+		}
+		r = r.WithContext(newCtx)
+		r.RemoteAddr = net.JoinHostPort(trusted.Host, trusted.Port)
+		h.ServeHTTP(w, r)
+		return
+	}), nil
+}
+
+// Addr represents only the Host and Port of a TCP address.
+type Addr struct {
+	Host string
+	Port string
+}
+
+// TrustedFromXForwardedFor will use the XForwardedFor* listener config settings
+// to determine how/if X-Forwarded-For are trusted/allowed for an inbound
+// request.  Important: return values of nil, nil are valid and simply means
+// that no "trusted" header was found and no error was raised as well.  Errors
+// can be raised for a number of conditions based on the listener config
+// settings, especially when the config setting for
+// XForwardedForRejectNotPresent is set to true which means if a "trusted"
+// header can't be found the request should be rejected.
+func TrustedFromXForwardedFor(r *http.Request, l *ListenerConfig) (*Addr, error) {
+	if r == nil {
+		return nil, fmt.Errorf("missing http request: %w", ErrInvalidParameter)
+	}
+	if l == nil {
+		return nil, fmt.Errorf("missing listener config: %w", ErrInvalidParameter)
+	}
+	rejectNotPresent := l.XForwardedForRejectNotPresent
+	hopSkips := l.XForwardedForHopSkips
+	authorizedAddrs := l.XForwardedForAuthorizedAddrs
+	rejectNotAuthz := l.XForwardedForRejectNotAuthorized
+
+	headers, headersOK := r.Header[textproto.CanonicalMIMEHeaderKey("X-Forwarded-For")]
+	if !headersOK || len(headers) == 0 {
+		if !rejectNotPresent {
+			return nil, nil
+		}
+		return nil, fmt.Errorf("missing x-forwarded-for header and configured to reject when not present")
+	}
+
+	host, port, err := net.SplitHostPort(r.RemoteAddr)
+	if err != nil {
+		// If not rejecting treat it like we just don't have a valid
+		// header because we can't do a comparison against an address we
+		// can't understand
+		if !rejectNotPresent {
+			return nil, nil
+		}
+		return nil, fmt.Errorf("error parsing client hostport: %w", err)
+	}
+
+	addr, err := sockaddr.NewIPAddr(host)
+	if err != nil {
+		// We treat this the same as the case above
+		if !rejectNotPresent {
+			return nil, nil
+		}
+		return nil, fmt.Errorf("error parsing client address: %w", err)
+	}
+
+	var found bool
+	for _, authz := range authorizedAddrs {
+		if authz.Contains(addr) {
+			found = true
+			break
+		}
+	}
+	if !found {
+		// If we didn't find it and aren't configured to reject, simply
+		// don't trust it
+		if !rejectNotAuthz {
+			return nil, nil
+		}
+		return nil, fmt.Errorf("client address not authorized for x-forwarded-for and configured to reject connection")
+	}
+
+	// At this point we have at least one value and it's authorized
+
+	// Split comma separated ones, which are common. This brings it in line
+	// to the multiple-header case.
+	var acc []string
+	for _, header := range headers {
+		vals := strings.Split(header, ",")
+		for _, v := range vals {
+			// validate the header contains a valid IP
+			v = strings.TrimSpace(v)
+			ip := net.ParseIP(v)
```

Comment:
In contrast to above, this I think will error if you _do_ have a port

Reply:
It doesn't error if there's no port: https://go.dev/play/p/jM1UsbsMXtH

Reply:
Right, but it does if you do have one! So the only way to get multiple addresses to be checked successfully would be for the first value to have a port and all subsequent values to not have ports. I doubt there's any proxy or LB that behaves that way.

I think in all cases we probably need to pass through `net.SplitHostPort` and either use the host value if a port was split or if we got an error about a missing port (I think it is sadly still not exported so it'd have to be a string check) we'd use the bare address.

Would love to get input from @briankassouf or @ncabatoff here.

Reply:
Ah.. I didn't realize that x-forwarded-for could have ports as well. ty!

Reply:
I don't know if they can or not. Maybe they never can. I was mixing up RemoteAddr above vs. X-Forwarded-For.

Reply:
I believe that it may have a port: https://www.rfc-editor.org/rfc/rfc7239#section-5.2

Reply:
That's Forwarded which is different -- an attempt to standardize the defacto X-Forwarded-For.

It won't _hurt_ if you want to detect port and use it if it's there. Then the replaced RemoteAddr can have a port too.

Reply:
We also stash the original port number and merge it back to the RemoteAddr, so there is another bit of code here assuming there will not be port number

Reply:
So... I think, we need to change how port is handled a bit.  This code and the current vault code, takes the trusted `acc` the remote address' port and assumes they belong together.  In vault this could result in the request remote address having a port that belongs to the inbound remote address but an IP that comes from a trusted x-forwarded-for header.

Reply:
Hum... the more I ponder this code, the more curious I get about port.  It would appear that a request's remote address must have a port... since x-forwarded-for headers aren't required to have a port, there seems to be a misalignment here when vault overwrites the remote address with the trusted header.

Reply:
@jefferai @briankassouf -> pushed a commit to address most of the comment I believe.. but still need to resolve how we want to handle port, especially when remote addr is overwritten with it.

Reply:
Did we ever figure out if X-Forwarded-For may include port numbers? Unfortunately it's not clear to me which port number we ought to use if there are two different ones. It probably makes sense to always ensure a port is set though so we don't break the assumption down the call chain that `RemoteAddr` always contains a port

Reply:
Based on some googling, It would appear that it's not uncommon it to have a port if folks are using IIS servers running ARR.

Reply:
Okay... pushed another commit which is a possible solution based on this discussion and it's ready for re-review. ty!